registry/warming.go

```diff
@@ -43,7 +42,7 @@ type backlogItem struct {
 func (w *Warmer) Loop(stop <-chan struct{}, wg *sync.WaitGroup, imagesToFetchFunc func() ImageCreds) {
 	defer wg.Done()
 
-	if w.Logger == nil || w.ClientFactory == nil || w.Expiry == 0 || w.Writer == nil || w.Reader == nil {
+	if w.Logger == nil || w.ClientFactory == nil || w.Expiry == 0 || w.Cache == nil {
```

Comment:
Yes, good point. Could move it. There are some things that are supplied _after_ the warmer is constructed, but these are either `nil`able or could be arguments. (or both)

Reply:
Can we just let it panic? It's not like we check for nil in the few hundred other places we can technically have it, and I assume this isn't a special case.

None of the fields checked can ever be nil in the current code, with the exception of `Cache` which may be nil iff the `memcached-hostname` flag is set to the empty string.
I think we have a mismatch between documented behaviour and actual behaviour there, though.

See: [flag definition](https://github.com/weaveworks/flux/blob/5514161b10412babd558d3d16842524d29915f53/cmd/fluxd/main.go#L90), but [no fallback defined](https://github.com/weaveworks/flux/blob/5514161b10412babd558d3d16842524d29915f53/cmd/fluxd/main.go#L239) for an empty hostname.

Reply:
> Can we just let it panic? It's not like we check for nil in the few hundred other places we can technically have it

True, but that's no reason to go further backwards. I would rather make the things that are mandatory into arguments (to a constructor or to this method, whichever). Then you have to go out of your way to supply a `nil` value, rather than just constructing an empty struct.

> [...] Cache which may be nil iff the memcached-hostname flag is set to the empty string

That's the real bug here, if so :/

Reply:
Well it's not a bug that was introduced here (is it?), so I reckon you can just open an issue and fix it separately from this PR.